### PR TITLE
16255 Continuation Out validation

### PR DESCRIPTION
**Pull request by vysakh-menon-aot**
*Issue #:* /bcgov/entity#16255

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the lear license (Apache 2.0).

### CODE REVIEW

File: legal-api/tests/unit/services/filings/validations/test_continuation_out.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+        identifier='BC1234567',
+        legal_type='BC',
+        state=Business.State.ACTIVE,
+        founding_date=datetime.utcnow(),
+        cco_expiry_date=datetime.utcnow()
+    )
+    filing = copy.deepcopy(FILING_HEADER)
+    filing['filing']['header']['name'] = 'continuationOut'
+
+    for country in pycountry.countries:
```
Comment (vysakh-menon-aot) [inline legal-api/tests/unit/services/filings/validations/test_continuation_out.py:124]: @severinbeauvais here is the validation for all valid jurisdiction as mentioned in schema

File: legal-api/src/legal_api/models/business.py [hunk 2]
```diff
@@ -190,13 +190,15 @@ class AssociationTypes(Enum):
...
+            'foreign_jurisdiction_region',
             'foreign_identifier',
             'foreign_legal_name',
             'foreign_legal_type',
             'foreign_incorporation_date',
             'send_ar_ind',
             'restoration_expiry_date',
-            'cco_expiry_date'
+            'cco_expiry_date',
+            'continuation_out_date'
```
Comment (severinbeauvais) [inline legal-api/src/legal_api/models/business.py:201]: Are these going away? I think UI doesn't need them.

File: legal-api/src/legal_api/services/filings/validations/continuation_out.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    return msg
+
+
+def validate_foreign_jurisdiction(filing: Dict, filing_type: str) -> list:
+    """Validate foreign jurisdiction."""
+    msg = []
+    foreign_jurisdiction_path = f'/filing/{filing_type}/foreignJurisdiction'
+
+    foreign_jurisdiction = filing['filing'][filing_type]['foreignJurisdiction']
+    country = pycountry.countries.search_fuzzy(foreign_jurisdiction['country'])[0].alpha_2
```
Comment (argush3) [inline legal-api/src/legal_api/services/filings/validations/continuation_out.py:80]: what happens if an invalid `country` is submitted in the filing json?

File: legal-api/src/legal_api/models/business.py (hunk 2 above)
Comment (vysakh-menon-aot) [reply legal-api/src/legal_api/models/business.py:201]: will remove `cco_expiry_date`

File: legal-api/src/legal_api/services/filings/validations/continuation_out.py (hunk 3 above)
Comment (vysakh-menon-aot) [reply legal-api/src/legal_api/services/filings/validations/continuation_out.py:80]: good catch. will add that validation

File: legal-api/src/legal_api/services/filings/validations/continuation_out.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    region = foreign_jurisdiction.get('region')
```
Comment (argush3) [inline legal-api/src/legal_api/services/filings/validations/continuation_out.py:81]: i think we'll need to deal with `undefined` or `null` scenarios for `region`

Comment (vysakh-menon-aot) [reply legal-api/src/legal_api/services/filings/validations/continuation_out.py:81]: you mean to make sure its `undefined` or `null` when country is other than `US ot CA`?

Comment (argush3) [reply legal-api/src/legal_api/services/filings/validations/continuation_out.py:81]: no, for `CA` and `US` scenarios.  

I think `if region == 'BC'` will be an issue. 

 Not sure if `elif not (region == 'FEDERAL' or pycountry.subdivisions.get(code=f'{country}-{region}')):` will error out.

Comment (severinbeauvais) [reply legal-api/src/legal_api/services/filings/validations/continuation_out.py:81]: > I think `if region == 'BC'` will be an issue.

There are lots of extra checks we could do but I agree this is a good one. (Even though our UI won't allow this region, the API is used by external users and should protect itself.)

Comment (argush3) [reply legal-api/src/legal_api/services/filings/validations/continuation_out.py:81]: actually, I think your code works.  Resolving.

File: legal-api/tests/unit/services/filings/validations/test_continuation_out.py [hunk 5]
```diff
@@ -72,6 +72,8 @@ def test_validate_continuation_out_date(session, test_name, expected_code, messa
 @pytest.mark.parametrize(
     'test_name, expected_code, message',
     [
+        ('FAIL_NO_COUNTRY', HTTPStatus.UNPROCESSABLE_ENTITY, None),
+        ('FAIL_INVAILD_COUNTRY', HTTPStatus.BAD_REQUEST, 'Invalid country.'),
```
Comment (severinbeauvais) [inline legal-api/tests/unit/services/filings/validations/test_continuation_out.py:76]: Typo.

**Review by severinbeauvais — APPROVED** [review]
LGTM

**Review by chenhongjing — APPROVED** [review]
:thumbsup:

**Pull request merged**